core/new-gui/src/app/workspace/service/autocomplete/model/autocomplete.service.ts

```diff
@@ -0,0 +1,185 @@
+
+    this.getSourceTableAddedOperatorMetadataObservable().subscribe(
+      metadata => { this.operatorSchemaList = metadata.operators; }
+    );
+
+    this.handleTexeraGraphPropertyChangeEvent();
+    this.handleTexeraGraphLinkChangeEvent();
+  }
+
+  /**
+   * This function returns the observable for operatorMetadata modified using the function addSourceTableNamesToMetadata. An important
+   * part to note here is that the function uses an combineLatest which fires first when it receives one value from each observable
+   * and thereafter a new value from any observable leads to refiring of the observables with the latest value fired by all the observables.
+   * Thus, any new value in either of the observables will cause recomputation of the modifed operator schema.
+   */
+  public getSourceTableAddedOperatorMetadataObservable(): Observable<OperatorMetadata> {
+    return  combineLatest(this.getSourceTablesNamesObservable(), this.operatorMetadataService.getOperatorMetadata())
+            .map(([tableNames, operatorMetadata]) =>
+            AutocompleteUtils.addSourceTableNamesToMetadata(operatorMetadata, tableNames));
+  }
+
+  /**
+   * Returns the observable which outputs a string everytime the autocomplete API is
+   * invoked and response is received successfully.
+   */
+  public getAutocompleteAPIExecutedStream(): Observable<string> {
+    return this.autocompleteAPIExecutedStream.asObservable();
+  }
+
+  /**
+   * Modifies the schema of the operator according to autocomplete information obtained from the backend (if autcomplete info
+   * contains the operator id).
+   * @param operator the operator whose property is to be displayed in the property editor
+   */
+  public findAutocompletedSchemaForOperator(operator: Readonly<OperatorPredicate>|undefined): OperatorSchema {
+    if (!operator) {
+      throw new Error(`autcomplete service:findAutocompletedSchemaForOperator - operator is undefined`);
+    }
+
+    const operatorSchema = this.operatorSchemaList.find(schema => schema.operatorType === operator.operatorType);
+    if (!operatorSchema) {
+      throw new Error(`operator schema for operator type ${operator.operatorType} doesn't exist`);
+    }
+
+    if (!(operator.operatorID in this.operatorInputSchemaMap)) {
+      return operatorSchema;
+    }
+
+    return AutocompleteUtils.addInputSchemaToOperatorSchema(operatorSchema, this.operatorInputSchemaMap[operator.operatorID]);
+  }
+
+  /**
+   * Used for automated propagation of input schema in workflow.
+   *
+   * When users are in the process of building a workflow, Texera can propagate schema forwards so
+   * that users can easily set the properties of the next operator. For eg: If there are two operators Source:Scan and KeywordSearch and
+   * a link is created between them, the attributed of the table selected in Source can be propagated to the KeywordSearch operator.
+   */
+  public invokeAutocompleteAPI(reloadCurrentOperatorSchema: boolean): void {
```

Comment:
I strongly feel having `reloadCurrentOperatorSchema` argument for this function is very wrongly designed. It doesn't feel right and breaks the single responsibility of this module. 

The logic is, depends on this `reloadCurrentOperatorSchema`, whether `autocompleteAPIExecutedStream` emits an event after autocomplete returns. Property editor updates its own operator schema when the event emits.

The `current operator`, which is the selected operator, only makes sense for the property panel, why does the autocomplete's behavior ever affected by some UI component? If in the future another component also needs autocomplete schema, and that component also has some certain behaviors, we would add another boolean variable for that. This is clearly not ideal.

Here's also another bug with the current design. If the operator is already filled with some values, then the schema is updated to be a drop-down list, the old value becomes invalid. Follow these steps to reproduce:
1. drag ScanSource and SentimentAnalysis. Don't connect the link yet.
2. Type some random things in the `Attribute` of SentimentAnalysis
3. Select a source table, like `twitter_sample` in ScanSource. Then connect the link.
4. Click Sentiment Analysis again, the following error will show:
<img width="195" alt="screen shot 2018-11-13 at 1 12 00 am" src="https://user-images.githubusercontent.com/12578068/48365239-54d5d800-e6e5-11e8-95ea-bd7fa8675f0e.png">


Here's a design I propose to both improve the current design and solve the bug.

The `operatorSchemaList` in `AutocompleteService` should be a single source of truth: it reflects the `dynamicSchema` of the current workflow state. `dynamicSchema` is changed dynamically because of changing workflow state and schema propagation. Each time the autocompleteService is invoked, it currently returns something like `Map<OperatorID, List<AttributeName>>`, then the `dynamicSchema` should be updated properly. This design also makes the separation of responsibilities more clear, `OperatorSchemaService` is also a single source of truth, it reflects the original `staticSchema` of the operators. 

`dynamicSchema` can be something like `Map<OperatorID(string), OperatorSchema>`. It could also just be the current `operatorInputSchemaMap`.

1. remove `reloadCurrentOperatorSchema` parameter. 

2. In `handleExecuteResult`, diff the new autocomplete result with the existing autocomplete result. For any changed operator input attributes, we need to change their schema `dynamicSchema`. 

3. For each changed schema operator, use a json schema validation library to check if the operator's property are still all valid. If a property is not valid anymore, remove the property of the operator in WorkflowGraph. 

4. Change `autocompleteAPIExecutedStream` to be `operatorSchemaChangedStream: Subject<string, OperatorSchema>`. For each operator schema that's changed, emit an event for each changed operator schema.

5. In PropertyEditor Component, listen to the new `operatorSchemaChangedStream` and see if the current operator's schema is changed. If it's changed, reload the operator property data from `workflowGraph`, and change the operator data and schema in the json schema form.

6. Since we now keep the `dynamicSchema` corresponds to each operator, we also needs to listen to operator add/delete stream. If an operator is added, add its original schema. If it's deleted, also delete it from the `dynamicSchemaMap`

Reply:
While I am implementing this new idea, I have some question regarding this new API. 

1. For this implementation, we have a `dynamicSchema` that keeps the new schema for all existing operators. When `handleExecuteResult` is called. We need to check if the new autocomplete result differs from the existing result. For this, do we need to generate the new incoming Schema by calling `findAutocompletedSchemaForOperator` and use `isEquals()` to check its difference from `dynamicSchema.get(operatorID)`? 
2. You mentioned that if the operator's property is invalid, we remove its property. Do we use another Json Schema library to check if the schema is valid, since the current library doesn't provide an API we can call to validate json schema unless we display it on the UI?
3. With `dynamicSchema`, when calling `findAutocompletedSchemaForOperator`, do we just get the schema by calling `dynamicSchema.get(operatorID)` since it already contains the newest schema?

Reply:
@henrychen0220 

2. Yes we probably need another library for the validation. I don't know if the current library we use exposes such API to validate the JSON schema against the data. We cannot rely on the "property editor component" to do this check. The reason is that the change in json schema might affect multiple operators. The property editor would only fix the data on one operator currently being displayed. Actually the current library might use other libraries to do the schema validation under the hood.

1 and 3. There are many ways to implement this diff, and you are not restricted to the function `findAutocompletedSchemaForOperator`. I would rather call this function `getDynamicSchema(operatorID: string): OperatorSchema`, and it could just contain a very simple `get` function. 

One possible way is what you described, every time we get the autocomplete result, we construct the new `OperatorSchema`. And then compare the new operator schema with the old operator schema as a whole. 

Another possible way is to also keep the last autocomplete result. Remember the autocomplete result is like a map from operatorID to the list of its input attributes. Such as `{"operator1": ["attr1", "attr2"], "operator2": ["attr2"] }`. We could also just diff the old autocomplete result with the new autocomplete result.

Note that regardless diffing what, we need to carefully pay attention to newly added / deleted operators.